pymchelper/readers/shieldhit.py

```diff
@@ -9,9 +9,364 @@
+            alanine_gy_units = ("Gy", "Alanine RE*Dose")
+
+        _detector_units = {
+            SHDetType.unknown: ("(nil)", "None"),
+            SHDetType.energy: ("MeV/primary", "Energy"),
+            SHDetType.fluence: (" cm^-2/primary", "Fluence"),
+            SHDetType.crossflu: (" cm^-2/primary", "Planar fluence"),
+            SHDetType.letflu: (" MeV/cm", "LET fluence"),
+            SHDetType.dose: dose_units,
+            SHDetType.dose_gy: dose_gy_units,
+            SHDetType.dlet: ("keV/um", "dose-averaged LET"),
+            SHDetType.tlet: ("keV/um", "track-averaged LET"),
+            SHDetType.avg_energy: ("MeV", "Average energy"),
+            SHDetType.avg_beta: ("(dimensionless)", "Average beta"),
+            SHDetType.material: ("(nil)", "Material number"),
+            SHDetType.alanine: alanine_units,
+            SHDetType.alanine_gy: alanine_gy_units,
+            SHDetType.counter: ("/primary", "Particle counter"),
+            SHDetType.pet: ("/primary", "PET isotopes"),
+            SHDetType.dletg: ("keV/um", "dose-averaged LET"),
+            SHDetType.tletg: ("keV/um", "track-averaged LET"),
+            SHDetType.zone: ("(dimensionless)", "Zone#"),
+            SHDetType.medium: ("(dimensionless)", "Medium#"),
+            SHDetType.rho: ("g/cm^3", "Density"),
+        }
+        return _detector_units.get(detector_type, ("(nil)", "(nil)"))
+
+
+class _SHBinaryReader0p6:
+    """
+    Binary format reader from version >= 0.6
+    """
+    def __init__(self, filename):
+        self.filename = filename
+
+    def read(self, detector, nscale=1):
+        with open(self.filename, "rb") as f:
+            d1 = np.dtype([('magic', 'S6'),
+                           ('end', 'S2'),
+                           ('vstr', 'S16')])
+
+            np.fromfile(f, dtype=d1, count=1)  # read the data into numpy
+            # print(x['magic'][0])
+            # print(x['end'][0])
+            # print(x['vstr'][0])
+
+            while(f):
+                token = self.get_token(f)
+                if token is None:
+                    break
+
+                pl_id, _pl_type, _pl_len, _pl = token
+
+                pl = [None]*_pl_len
+
+                # print("_pl_type",_pl_type.decode('ASCII'))
+                # decode all strings (currently there will never be more than one per token)
+                if 'S' in _pl_type.decode('ASCII'):
+                    for i, _j in enumerate(_pl):
+                        pl[i] = _pl[i].decode('ASCII')
+                else:
+                    pl = _pl
+
+                # print("0x{:02x}".format(pl_id))
+
+                # TODO: some clever mapping could be done here surely
+                # something like this: however the keymaps are not complete
+                # attr_keys = SHBDOTagID(pl_id)
+                # attributes = mapping[attr_keys]
+                # for i, attr in enumerate(attributes):
+                #     setattr(detector,attr,pl[i])
```

Comment:
decided not to do this in this step, as handling all the exceptions made the code rather complicated.
Refactoring this would be a new issue.